### ISSUE TITLE
Can you link in the c library?

### ISSUE THREAD

**Issue by Cat22** [davidcallanan/os-series#25]
If you could link in libc (/lib64/libc.so.6) then you could use standard C functions like printf. I'll bet this isnt all that difficult and you could look at startup code for an ordinary C program to get the method/code/info on how to do it. 
Plus it would make a great video!

**Comment by Kjur0** [comment]
You have to add
`#include <stdio.h>`
on the first line of your .c file...
But it isn't working for me

**Comment by tayoky** [comment]
there are some littles things you have to understand : userspace apps and kernel
## userspace apps
userspace apps are normal application they can use library like the c library , include stdio.h and all that stuff
## kernel
kernel cannot use any function of any library so you cannot use stdio.h  because stdio (and all other libc functions) are in userpsace ln top of the kernel
kernel must provide everything  
## the libc
the libc is build like a normal libary on top of the kernel so to make stdio.h you need to have a advanced kernel with userspace and some syscall 

if you need exemple take a look at my [my libc](https://github.com/tayoky/tlibc) and [my os](https://github.com/tayoky/stanix)

**Comment by davidcallanan** [comment]
It is not expected that the C standard for library would work, as it depends on a fully functioning operating system to work.

You may re-implement some functionality as needed. It is unlikely that I would ever support the full C standard library, but I'll keep this open for now.

**Comment by tayoky** [comment]
> It is not expected that the C standard for library would work, as it depends on a fully functioning operating system to work.
> 
> You may re-implement some functionality as needed. It is unlikely that I would ever support the full C standard library, but I'll keep this open for now.

@davidcallanan it's not that hard
i've been working on my own os for about 6 month and i aready have most of the c standard library done

**Comment by davidcallanan** [comment]
@tayoky TBH you're right, there's not as much in it as I thought. I thought threads were part of it, but really it's just memory allocation which is trivial. Now, I might reconsider. Except filesystem is needed for some functionality, although nothing stops me from implementing that in-memory. 😂

**Comment by tayoky** [comment]
well there a bit more @davidcallanan if you want to be posix compilant exec/fork (so multitasking), signal, ....

**Comment by tayoky** [comment]
@davidcallanan but before any c library there should be a vfs program loading and userspace (that the order)

**Comment by davidcallanan** [comment]
@tayoky Yeah, it would take a lot of work to get to being POSIX-compliant.

**Comment by tayoky** [comment]
@davidcallanan but it's 100% possible and i managed to do it (and this mean i can run third party program such as TCC or doom)

**Comment by davidcallanan** [comment]
@tayoky Wow, congrats! Curious have you attempted to implement networking?

**Comment by tayoky** [comment]
> @tayoky Wow, congrats! Curious have you attempted to implement networking?

@davidcallanan nop but i have some idea on how to do it

**Comment by davidcallanan** [comment]
@tayoky It's inconvenient because there's no universal networking standard. I was looking at implementing it for AWS, it's like 17000 LOC, but it's open source so can probably steal most of it. (That only covers the base networking layer, not UDP etc).

**Comment by tayoky** [comment]
> @tayoky It's inconvenient because there's no universal networking standard. I was looking at implementing it for AWS, it's like 17000 LOC, but it's open source so can probably steal most of it. (That only covers the base networking layer, not UDP etc).

@davidcallanan i'm not offending you but you should probably read a bit of theory about networking
the driver for the network card (NiC)
should not implement udp or tpc just allow to send and recive eterneth packet
on top of that you have another driver the ipv4 driver
this driver can send ip request using a NiC driver but it don't care about what card is used so it's generic
on top of that you have the udp/tcp driver that use the ipv4 driver to send udp/tcp request (on some os the ipv4 and udp/tcp is merged in one but NEVER the NiC driver and the ipv4 should be merged)
also stealing driver is realllllllyyy hard as your kernel will probably won't work the exact same way 
the c library (in userspace) can imolement the pre last layer by doing dns request using the udp driver to turn domain name into ip
then the application can use the tcp driver to do https request for exemple

**Comment by davidcallanan** [comment]
@tayoky Yes I am aware of how to layer the drivers, that's why I mentioned the base layer, which is the most inconvenient because of the differences between devices. The rest of the layers are not so inconvenient, just tedious. So for the base driver 17,000 LOC at 100 lines per day, that's 170 days. Sure there'd be a lot of differences with my kernel, but probably at least half that code would remain unchanged. If I theoretically ever did this, I would probably implement everything beyond the ethernet layer in userspace.

**Comment by tayoky** [comment]
> @tayoky Yes I am aware of how to layer the drivers, that's why I mentioned the base layer, which is the most inconvenient because of the differences between devices. The rest of the layers are not so inconvenient, just tedious. So for the base driver 17,000 LOC at 100 lines per day, that's 170 days. Sure there'd be a lot of differences with my kernel, but probably at least half that code would remain unchanged. If I theoretically ever did this, I would probably implement everything beyond the ethernet layer in userspace.

@davidcallanan sorry i mistunderstood your message
you can start with something simpler for realtek NiC and e1000 card a 
simple driver can be between 600 to 1000 line of code
(btw i know some people doing tcp driver it's really hard)

**Comment by Cat22** [comment]
Where I worked before the bios writers had managed to switch from asm to C. I don't now much else about that but it greatly improved their code and made debug a lot easier

**Comment by tayoky** [comment]
> Where I worked before the bios writers had managed to switch from asm to C. I don't now much else about that but it greatly improved their code and made debug a lot easier

@Cat22 whay do you mean ?

**Comment by tayoky** [comment]
@davidcallanan i can help on the serie if you want

**Comment by davidcallanan** [comment]
> [@davidcallanan](https://github.com/davidcallanan) sorry i mistunderstood your message you can start with something simpler for realtek NiC and e1000 card a simple driver can be between 600 to 1000 line of code (btw i know some people doing tcp driver it's really hard)

@tayoky That's good to know, thanks. Tbh it will be at least another 5 months before I ever look at this again.
